test/unit/eigensolver/test_reduction_to_band.cpp

```diff
@@ -188,74 +191,66 @@ void splitReflectorsAndBand(MatrixLocal<const T>& mat_v, MatrixLocal<T>& mat_b,
+    }
   }
 
+  DLAF_ASSERT(to_SizeType(taus.size()) == k, taus.size(), k);
+
   return taus;
 }
 
 template <class T>
-auto allGatherTaus(const SizeType k, const SizeType chunk_size,
-                   std::vector<pika::shared_future<common::internal::vector<T>>> fut_local_taus,
+auto allGatherTaus(const SizeType k, Matrix<T, Device::CPU>& mat_taus,
                    comm::CommunicatorGrid comm_grid) {
   std::vector<T> taus;
   taus.reserve(to_sizet(k));
 
-  pika::wait_all(fut_local_taus);
-  auto local_taus = pika::unwrap(fut_local_taus);
-
-  const SizeType n_chunks = dlaf::util::ceilDiv(k, chunk_size);
-
-  // Note:
-  // this is just an early exit
-  const SizeType n_local_chunks =
-      n_chunks / comm_grid.size().cols() +
-      (comm_grid.rank().col() < (n_chunks % comm_grid.size().cols()) ? 1 : 0);
-
-  if (local_taus.size() != to_sizet(n_local_chunks))
-    return taus;
-
-  for (auto index_chunk = 0; index_chunk < n_chunks; ++index_chunk) {
-    const auto owner = index_chunk % comm_grid.size().cols();
+  for (SizeType i = 0; i < mat_taus.nrTiles().cols(); ++i) {
+    const auto owner = mat_taus.rankGlobalTile(GlobalTileIndex(0, i)).col();
     const bool is_owner = owner == comm_grid.rank().col();
 
-    const auto this_chunk_size = std::min(k - index_chunk * chunk_size, chunk_size);
+    const auto chunk_size = mat_taus.tileSize(GlobalTileIndex(0, i)).cols();
 
-    std::vector<T> chunk_data;
     if (is_owner) {
-      const auto index_chunk_local = to_sizet(index_chunk / comm_grid.size().cols());
-      chunk_data = local_taus.at(index_chunk_local);
-      sync::broadcast::send(comm_grid.rowCommunicator(),
-                            common::make_data(chunk_data.data(),
-                                              static_cast<SizeType>(chunk_data.size())));
+      auto tile_local = sync_wait(mat_taus.read(GlobalTileIndex(0, i)));
+      sync::broadcast::send(comm_grid.rowCommunicator(), common::make_data(tile_local.get()));
+      for (SizeType k = 0; k < tile_local.get().size().cols(); ++k) {
+        taus.push_back(tile_local.get()(TileElementIndex(0, k)));
+      }
     }
     else {
-      chunk_data.resize(to_sizet(this_chunk_size));
-      sync::broadcast::receive_from(owner, comm_grid.rowCommunicator(),
-                                    common::make_data(chunk_data.data(),
-                                                      static_cast<SizeType>(chunk_data.size())));
+      Tile<T, Device::CPU> tile_local(TileElementSize(1, chunk_size),
+                                      MemoryView<T, Device::CPU>(chunk_size), 1);
+      sync::broadcast::receive_from(owner, comm_grid.rowCommunicator(), common::make_data(tile_local));
+      for (SizeType k = 0; k < chunk_size; ++k) {
+        taus.push_back(tile_local(TileElementIndex(0, k)));
+      }
     }
 
     // copy each chunk contiguously
-    std::copy(chunk_data.begin(), chunk_data.end(), std::back_inserter(taus));
   }
 
+  // TODO: is the early exit required?
```

Comment:
@albestro, @rasolca there was an early exit here (https://github.com/eth-cscs/DLA-Future/blob/34934c60c5cc6f0c1b71fc5e2148e2c89b6903d1/test/unit/eigensolver/test_reduction_to_band.cpp#L225-L232 on master). The comment says "This is just and early exit" but it's not quite clear to me what the purpose of the early exit is. Is it optional? If the condition isn't true it seems like the whole test should fail. Should it be an `ASSERT*` or `TEST*` instead?

Reply:
Not sure. @albestro?
Anyway, is this routine still needed? Can you replace it with one of the matrix allgather methods available in `dlaf_test` or it doesn't work as the distribution is "hand made"?

Reply:
I hadn't thought about using `allGather`. In the distributed case it fails on https://github.com/eth-cscs/DLA-Future/blob/2634d5bde89efc80b18b76b4afaeefb65618a1e6/test/include/dlaf_test/matrix/util_matrix_local.h#L104, understandably. Am I missing some other helper that you think could help? If not, I would probably leave this as it is given that the plan is to clean this up once we have a distributed vector. What do you think?

Reply:
Back to the original question:
You can use either `DLAF_ASSERT` or `EXPECT_EQ`.
`ASSERT_EQ` cannot be used in non-void functions.

Reply:
I added back a check (as `EXPECT_EQ`) in https://github.com/eth-cscs/DLA-Future/pull/908/commits/f856457f4dba1ec0046fb704f703096eadc9248f for the distributed case. Not sure it adds much, but probably doesn't hurt for now. We can always remove it if seems superfluous or annoying in the future.

Reply:
I admit my comment is not helpful at all.

Not sure why this check is there...after a quick view I was not able to fully reconstruct the history of this change. It might be either an additional check I added during development or a leftover. It is something marginal, so I would not bother too much. Anyway, your adaptation looks good 😉